### ISSUE TITLE
some Questions

### ISSUE THREAD

**Issue by CCV-Edward** [Adopteruf/Action_Recognition_using_Visual_Attention#1]
Thanks for your wonderful work!

I get some questions about your work,
1, why not combine 'build_model' and 'build_sampler' together?  'loss' and 'sampled_label_index_list' both are generated by logits, if so, it would reduce many parameters.
2, 'models = sorted(models, cmp=model_comp)' in solve.py, it cannot sort all model in order. (another lower-level question: which file would be restored in model)
3, I want to transform this repository in UCF11, but result cannot be converged, please give me some advice.

regards,
Edward.

**Comment by jingweio** [comment]
I am sorry about reading this question so late!!

1. Actually, the whole structure is what I adopted from the source code of the image caption ["Show, Attend and Tell: Neural Image Caption Generation with Visual Attention"] as referenced in README. It is true that combining both methods can save lots of variables. But, personally speaking, I would still prefer to separate those two functions for the flexibility and pertinence it offers. In this way, I can debug more easily. During the training process, I can use 'build_sampler' to directly show whether the current condition of the model based on training set or I can just shutdown 'build_sampler' and see the change of cost to judge whether the model is converging or not. During the testing process, I can only use 'build_sampler' to evaluate the permanence of the trained model and even print out some useful parameters such as the coefficients assigned by attention model to find where the model is currently focusing on, which is not necessary during the training process. Anyway, it is just matter of personal habit. There is no need to worry about too much.
2. By writing this line of code, I want to order the trained model by the number of iteration spent on them for the following visualization. In order to achieve that, I build a function 'model_comp' in utils.py to realize the specific purpose. It should work. If it does not work, I guess it can be solved by debugging the codes a little bit to see what actually happen inside. Besides, the models will be saved under .../data/model/ or some where you set by yourself in the beginning.
3. Except for the issues concerning the codes, I would like to suggest you to increase number of LSTM cells, increase the layer of LSTM(there are only one layer of LSTM in this case), change the feature extraction CNN to ResNet50(which is very popular now) or use the HM-LSTM model instead of the usual LSTM model.

regards,
Adopteruf

**Comment by CCV-Edward** [comment]
Thanks for your attention! I have solved all the problems and thank you very much!

Besides, i wanna know if you have ResNet50 model just like vgg19.mat, which can be transferred into your scripts. Additional, could you please explain HM-LSTM or give me some sites to learn for this unknown model. In my work, i want to show the 'Confusion Matrix', can you give me some idea to achieve it.

**Comment by jingweio** [comment]
I would like to recommend you to import "Keras" using tensorflow as BK to bring ResNet50 to you scripts. Moreover, you can first download weights of the pre-trained model on ImageNet online manually and then just load the weights(.h document) into to structure of ResNet50 provided by "Keras". The difference in HM-LSTM is that it offers the chance to find the extreme conditions including "do not forget anything" and "forget everything". 
You can learn this model from the paper-"HIERARCHICAL MULTISCALE RECURRENT NEURAL NETWORKS arXiv:1609.01704v7 [cs.LG] 9 Mar 2017".
As for "Confusion Matrix", you can try "tensorboard" to find a way out.

**Comment by CCV-Edward** [comment]
could you please leave your email address?
i have some ideas, and i want to work with you together @Adopteruf

**Comment by jingweio** [comment]
I am sorry that I didn't notice your words util now for the couple busy weeks on the final exam. 
Sure, I'd love to learn about your ideas. I am currently working on the fields concerning re-identification which is also the center of my final year project. By the way, I am a year 4 student. But I still have interest in  action detection and image caption.
My email is adopteruf@outlook.com
I am looking forward to your reply.
